catkit2/services/oceanoptics_spectrometer/oceanoptics_spectrometer.py

```diff
@@ -0,0 +1,189 @@
+        The main function of the service.
+    close()
+        Close the service.
+    take_one_spectrum()
+        Get a spectrum from the spectrometer.
+    '''
+    NUM_FRAMES = 20
+
+    def __init__(self):
+        '''
+        Create a new OceanOpticsSpectrometer service.
+        '''
+        super().__init__('oceanoptics_spectrometer')
+
+        self.serial_number = str(self.config['serial_number'])
+        self.interval = self.config.get('interval', 1)
+        self._exposure_time = self.config.get('exposure_time', 1000)
+
+        self.spectrometer = None
+        self.model = None
+        self.pixels_number = None
+
+    def open(self):
+        '''
+        Open the service.
+
+        This function is called when the service is opened.
+        It initializes the spectrometer and creates the data streams and properties.
+
+        Raises
+        ------
+        ImportError
+            If the spectrometer cannot be found.
+        '''
+
+        # Find the spectrometer
+        try:
+            self.spectrometer = Spectrometer.from_serial_number(self.serial_number)
+        except SeaBreezeError:
+            raise ImportError(f'OceanOptics: Could not find spectrometer with serial number {self.serial_number}')
+
+        # exctract attributes
+        self.model = self.spectrometer.model
+        self.pixels_number = self.spectrometer.pixels
+
+        # Define and set default exposure time.
+        self.exposure_time = self._exposure_time
+
+        # Create datastreams
+        self.spectra = self.make_data_stream('spectra', 'float32', [self.pixels_number], self.NUM_FRAMES)
+
+        self.is_saturating = self.make_data_stream('is_saturating', 'int8', [1], self.NUM_FRAMES)
+        self.is_saturating.submit_data(np.array([0], dtype='int8'))
+
+        # Create properties
+        def make_property_helper(name, read_only=False):
+            if read_only:
+                self.make_property(name, lambda: getattr(self, name))
+            else:
+                self.make_property(name, lambda: getattr(self, name), lambda val: setattr(self, name, val))
+
+        make_property_helper('exposure_time')
+        make_property_helper('wavelengths', read_only=True)
+
+    def main(self):
+        '''
+        The main function of the service.
+
+        This function is called when the service is started.
+        '''
+        while not self.should_shut_down:
+            intensities = self.take_one_spectrum()
+            self.spectra.submit_data(np.array(intensities, dtype='float32'))
+            self.sleep(self.interval)
+
+    def take_one_spectrum(self):
+        '''
+        Measure one spectrum and submit it to the data stream.
+        '''
+        intensities = self.spectrometer.intensities()
+
+        if np.max(intensities) ==  self.spectrometer.max_intensity:
+            self.is_saturating.submit_data(np.array([1], dtype='int8'))
+        else:
+            self.is_saturating.submit_data(np.array([0], dtype='int8'))
```

Comment:
I have been thinking about it, and I think it might be better if this function really just measures and returns the intensities and to move the saturation check and its data stream submission to the main method too. In this way, the `take_one_spectrum()` can be regarded as a function that does exactly that. It would be a bit cleaner.

Reply:
done